File: csharp/tests/TestSerializingIntegration.cs
using MyAirtable;
using Xunit;

namespace MyAirtable.Tests;

/// <summary>
/// CS10.1 — Network-bound serialization cases (split from the offline <see cref="TestSerializing"/>):
/// id + createdTime round-trip after fetch, linked-record fields serializing as record-ID arrays,
/// and an explicit null clearing a writable field server-side. Parity with the Java/Kotlin suites.
/// </summary>
public class TestSerializingIntegration
{
    private readonly Airtable _airtable = TestSetup.MakeAirtable();

    [Fact]
    public async Task RoundTripPreservesIdAndCreatedTimeAfterFetch()
    {
        var primaryKey = TestSetup.PrimaryKey("Serializing", "IdTime");
        var created = await _airtable.Primary.Orm.CreateAsync(
            new PrimaryModel { PrimaryKey = primaryKey }
        );
        var recordId = created.Id!;
        try
        {
            Assert.NotNull(created.CreatedTime); // populated on create
            var fetched = await _airtable.Primary.Orm.GetAsync(recordId);
            Assert.Equal(recordId, fetched.Id);
            Assert.Equal(created.CreatedTime, fetched.CreatedTime);
        }
        finally
        {
            await TryDeletePrimaryAsync(recordId);
        }
    }

    [Fact]
    public async Task LinkedRecordFieldsSerializeAsRecordIdArrays()
    {
        var suite = TestSetup.PrimaryKey("Serializing", "Links");
        string? secId = null;
        string? primId = null;
        try
        {
            var sec = await _airtable.Secondary.Orm.CreateAsync(
                new SecondaryModel { Name = suite + " Target" }
            );
            secId = sec.Id;
            var prim = await _airtable.Primary.Orm.CreateAsync(
                new PrimaryModel
                {
                    PrimaryKey = suite,
                    LinkSingle = new List<string> { secId! },
                }
            );
            primId = prim.Id;

            var record = prim.ToRecord();
            var linkElement = record[PrimaryFields.LinkSingleId];
            Assert.NotNull(linkElement);
            var array = linkElement!.AsArray(); // links serialize as arrays
            Assert.Equal(secId, array[0]!.GetValue<string>());
        }
        finally
        {
            await TryDeletePrimaryAsync(primId);
            await TryDeleteSecondaryAsync(secId);
        }
    }

    [Fact]
    public async Task ExplicitNullOnAWritableFieldClearsItServerSide()
    {
        var primaryKey = TestSetup.PrimaryKey("Serializing", "Clear");
        var created = await _airtable.Primary.Orm.CreateAsync(
            new PrimaryModel { PrimaryKey = primaryKey, SingleLineText = "to be cleared" }
        );
        var recordId = created.Id!;
        try
        {
            created.SingleLineText = null;
            var saved = await _airtable.Primary.Orm.UpdateAsync(created);
            Assert.Null(saved.SingleLineText); // null dirty entry clears the field

            var fetched = await _airtable.Primary.Orm.GetAsync(recordId);
            Assert.Null(fetched.SingleLineText);
        }
        finally
        {
            await TryDeletePrimaryAsync(recordId);
        }
    }

    private async Task TryDeletePrimaryAsync(string? id)
    {
        if (string.IsNullOrEmpty(id))
            return;
        try
        {
            await _airtable.Primary.Orm.DeleteAsync(id);
        }
        catch (AirtableException) { }
    }

    private async Task TryDeleteSecondaryAsync(string? id)
    {
        if (string.IsNullOrEmpty(id))
            return;
        try
        {
            await _airtable.Secondary.Orm.DeleteAsync(id);
        }
        catch (AirtableException) { }
    }
}

File: cpp/tests/test_serializing_integration.cpp
// Network-bound serialization cases, split from the offline test_serializing
// suite (C# TestSerializingIntegration parity, 3 cases): id + createdTime
// round-trip after fetch, linked-record fields serializing as record-ID
// arrays, and an explicit null clearing a writable field server-side.
#include <catch2/catch_test_macros.hpp>

#include <string>
#include <vector>

#include "test_setup.hpp"

using namespace myairtable;
using myairtable_tests::make_airtable;
using myairtable_tests::primary_key;

namespace {

void try_delete_primary(Airtable& airtable, const std::string& id) {
    if (id.empty()) {
        return;
    }
    try {
        airtable.primary().delete_one(id);
    } catch (const AirtableException&) {
    }
}

void try_delete_secondary(Airtable& airtable, const std::string& id) {
    if (id.empty()) {
        return;
    }
    try {
        airtable.secondary().delete_one(id);
    } catch (const AirtableException&) {
    }
}

} // namespace

TEST_CASE("serializing integration: round trip preserves id and created time after fetch",
          "[json][serializing-integration]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Serializing", "IdTime");
    auto created = airtable.primary().create_one(PrimaryModel{.primary_key = pk});
    const auto record_id = *created.id;
    try {
        REQUIRE(created.created_time.has_value()); // populated on create
        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.id == record_id);
        REQUIRE(fetched.created_time == created.created_time);
    } catch (...) {
        try_delete_primary(airtable, record_id);
        throw;
    }
    try_delete_primary(airtable, record_id);
}

TEST_CASE("serializing integration: linked record fields serialize as record id arrays",
          "[json][serializing-integration]") {
    auto airtable = make_airtable();
    const auto suite = primary_key("Serializing", "Links");
    std::string sec_id;
    std::string prim_id;
    try {
        auto sec = airtable.secondary().create_one(SecondaryModel{.name = suite + " Target"});
        sec_id = *sec.id;
        auto prim = airtable.primary().create_one(
            PrimaryModel{.link_single = std::vector<std::string>{sec_id}, .primary_key = suite});
        prim_id = *prim.id;

        const auto record = prim.to_record();
        REQUIRE(record.contains(std::string(PrimaryFields::kLinkSingleId)));
        const auto& link_element = record.at(std::string(PrimaryFields::kLinkSingleId));
        REQUIRE(link_element.is_array()); // links serialize as arrays
        REQUIRE(link_element[0].get<std::string>() == sec_id);
    } catch (...) {
        try_delete_primary(airtable, prim_id);
        try_delete_secondary(airtable, sec_id);
        throw;
    }
    try_delete_primary(airtable, prim_id);
    try_delete_secondary(airtable, sec_id);
}

TEST_CASE("serializing integration: explicit null on a writable field clears it server side",
          "[json][serializing-integration]") {
    auto airtable = make_airtable();
    const auto pk = primary_key("Serializing", "Clear");
    auto created = airtable.primary().create_one(
        PrimaryModel{.primary_key = pk, .single_line_text = "to be cleared"});
    const auto record_id = *created.id;
    try {
        created.single_line_text = std::nullopt;
        auto saved = airtable.primary().update_one(created);
        REQUIRE(saved.single_line_text == std::nullopt); // null dirty entry clears the field

        auto fetched = airtable.primary().get_one(record_id);
        REQUIRE(fetched.single_line_text == std::nullopt);
    } catch (...) {
        try_delete_primary(airtable, record_id);
        throw;
    }
    try_delete_primary(airtable, record_id);
}
